File: create-fusee-go/internal/assets/embed/framework/core/compiler/main-compiler.js
import { compile as rustCompile } from './rust-compiler.js'
import { tokenize } from './lexer.js'
import { parse } from './parser.js'
import { transform } from './transformer.js'
import { CompileError } from './errors.js'

export function compile(source, options = {}) {
    const filename = options.filename ?? '<template>'
    const components = new Set(options.components ?? [])
    const target = options.target ?? 'client'

    try {
        const result = rustCompile(source, {
            filename,
            runtimePath: options.runtimePath,
            ssrRuntimePath: options.ssrRuntimePath ?? options.runtimePath,
            components: [...components],
            scope: options.scope ? [...options.scope] : [],
            throwOnWarning: options.throwOnWarning ?? false,
            target,
        })
        return {
            code: result.code,
            ssrCode: result.ssrCode ?? null,
            ast: result.ast,
            tokens: result.tokens,
            warnings: result.warnings ?? [],
        }
    } catch (err) {
        _rethrow(err, filename)
    }
}

export function parseOnly(source, options = {}) {
    const components = new Set(options.components ?? [])
    const tokens = tokenize(source)
    const ast = parse(tokens, source, components)
    return { ast, tokens }
}

export function transformOnly(ast, options = {}) {
    return transform(ast, {
        components: new Set(options.components ?? []),
        source: options.source ?? '',
        scope: options.scope,
    })
}

export function compileBatch(templates, sharedOptions = {}) {
    return templates.map(({ id, source }) => {
        try {
            const { code, ssrCode, warnings } = compile(source, { ...sharedOptions, filename: id })
            return { id, code, ssrCode, warnings, error: null }
        } catch (err) {
            return { id, code: null, ssrCode: null, warnings: [], error: err }
        }
    })
}

export function fuseePlugin(pluginOptions = {}) {
    return {
        name: 'vite-plugin-fusee',

        transform(src, id, ssrOptions) {
            if (!id.endsWith('.fusee') && !id.endsWith('.fhtml') && !id.endsWith('.template.html')) return null

            const isSSR = typeof ssrOptions === 'boolean' ? ssrOptions : Boolean(ssrOptions?.ssr)
            const { code, ssrCode, warnings } = compile(src, {
                filename: id,
                target: isSSR ? 'ssr' : 'both',
                runtimePath: pluginOptions.runtimePath,
                ssrRuntimePath: pluginOptions.ssrRuntimePath,
                components: pluginOptions.components,
                scope: pluginOptions.scope,
            })

            for (const w of warnings) {
                this.warn(typeof w.format === 'function' ? w.format(id) : String(w))
            }

            const outputCode = isSSR ? (ssrCode ?? code) : code
            return { code: outputCode, map: null }
        }
    }
}

export { fileRouterPlugin } from './plugins/file-router-plugin.js'
export { actionsPlugin } from './plugins/actions-plugin.js'
export { routerPlugin } from './plugins/router-plugin.js'
export { compileFileRoutes, generateRoutesModule, generateRouteTypes, validateRoutes as validateFileRoutes } from './plugins/file-router-plugin.js'
export { validateActions, generateClientStubs, generateServerRoutes, generateActionTypes } from './plugins/actions-plugin.js'
export { validateRoutes, generateRouterTypes } from './plugins/router-plugin.js'

function _rethrow(err, filename) {
    if (err instanceof CompileError) {
        err.message = err.format(filename)
        throw err
    }
    throw err
}

File: create-fusee-go/list-embed/framework/core/rust-compiler/src/main_compiler.rs
use crate::ast::Node;
use crate::errors::{CompileError, CompileWarning};
use crate::generator::{generate, GenerateOptions};
use crate::lexer::{tokenize, Token};
use crate::parser::parse;
use crate::ssr_generator::{generate_ssr, SSRGenerateOptions};
use crate::transformer::{transform, TransformOptions};
use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompileTarget {
    Client,
    SSR,
    Both,
}

impl Default for CompileTarget {
    fn default() -> Self {
        CompileTarget::Client
    }
}

pub struct CompileOptions {
    pub filename: String,
    pub components: HashSet<String>,
    pub scope: HashSet<String>,
    pub runtime_path: Option<String>,
    pub ssr_runtime_path: Option<String>,
    pub throw_on_warning: bool,
    pub target: CompileTarget,
}

impl Default for CompileOptions {
    fn default() -> Self {
        CompileOptions {
            filename: "<template>".to_string(),
            components: HashSet::new(),
            scope: HashSet::new(),
            runtime_path: None,
            ssr_runtime_path: None,
            throw_on_warning: false,
            target: CompileTarget::Client,
        }
    }
}

pub struct CompileResult {
    pub code: String,
    pub ssr_code: Option<String>,
    pub ast: Node,
    pub tokens: Vec<Token>,
    pub warnings: Vec<CompileWarning>,
}

pub fn compile(source: &str, options: CompileOptions) -> Result<CompileResult, CompileError> {
    let tokens = tokenize(source).map_err(|e| rethrow(e, &options.filename))?;

    let ast = parse(tokens.clone(), source, &options.components)
        .map_err(|e| rethrow(e, &options.filename))?;

    let transform_opts = TransformOptions {
        components: options.components.clone(),
        source: source.to_string(),
        scope: options.scope.clone(),
    };
    let (ast, warnings) = transform(ast, transform_opts);

    if options.throw_on_warning {
        if let Some(w) = warnings.first() {
            return Err(CompileError::new(w.code, w.loc.clone(), source, &[]));
        }
    }

    let root = match &ast {
        Node::Root(r) => r,
        _ => unreachable!("parse() always produces a Root node"),
    };

    let code: String;
    let mut ssr_code: Option<String> = None;

    match options.target {
        CompileTarget::Client => {
            code = generate(
                root,
                GenerateOptions {
                    source: source.to_string(),
                    runtime_path: options.runtime_path.clone(),
                },
            )
            .map_err(|e| rethrow(e, &options.filename))?;
        }
        CompileTarget::SSR => {
            let ssr = generate_ssr(
                root,
                SSRGenerateOptions {
                    source: source.to_string(),
                    runtime_path: options
                        .ssr_runtime_path
                        .clone()
                        .or_else(|| options.runtime_path.clone()),
                },
            )
            .map_err(|e| rethrow(e, &options.filename))?;
            code = ssr.clone();
            ssr_code = Some(ssr);
        }
        CompileTarget::Both => {
            code = generate(
                root,
                GenerateOptions {
                    source: source.to_string(),
                    runtime_path: options.runtime_path.clone(),
                },
            )
            .map_err(|e| rethrow(e, &options.filename))?;

            let ssr = generate_ssr(
                root,
                SSRGenerateOptions {
                    source: source.to_string(),
                    runtime_path: options
                        .ssr_runtime_path
                        .clone()
                        .or_else(|| options.runtime_path.clone()),
                },
            )
            .map_err(|e| rethrow(e, &options.filename))?;
            ssr_code = Some(ssr);
        }
    }

    Ok(CompileResult {
        code,
        ssr_code,
        ast,
        tokens,
        warnings,
    })
}

pub fn parse_only(source: &str, components: &HashSet<String>) -> Result<(Node, Vec<Token>), CompileError> {
    let tokens = tokenize(source)?;
    let ast = parse(tokens.clone(), source, components)?;
    Ok((ast, tokens))
}

pub fn transform_only(ast: Node, components: HashSet<String>, source: String, scope: HashSet<String>) -> (Node, Vec<CompileWarning>) {
    transform(
        ast,
        TransformOptions {
            components,
            source,
            scope,
        },
    )
}

fn rethrow(err: CompileError, filename: &str) -> CompileError {
    let formatted = err.format(filename);
    CompileError {
        message: formatted,
        ..err
    }
}
